### ISSUE TITLE
Deadlock in the pipeline on Exception?

### ISSUE THREAD

**Issue by kanadaj** [aspnet/AspNetCore#3887]
I've been trying to hunt down this issue for a while when our web application suddenly locks up with 100% CPU usage. I've finally acquired a full memory dump and analysis (fairly hard when the CPU is panicking).

It appears that somehow Exception.RestoreExceptionDispatchInfo is somehow deadlocking itself?

![image](https://user-images.githubusercontent.com/3247357/48036792-c354ec00-e161-11e8-9dcc-09ecb389cab1.png)

This is using .NET Framework 4.7.1. I'm not quite sure if this is related to ASP.NET Core or just a bug in the framework overall, or if this is possible at all...

EDIT: Opening the dump, the exception hanging in thread 151 is `An existing connection was forcibly closed by the remote host`

**Comment by Tratcher** [comment]
Likely a duplicate of https://github.com/aspnet/KestrelHttpServer/issues/2806. Do you have a reliable repro? Can you try it on 2.2.0-preview2 or later?

**Comment by kanadaj** [comment]
@Tratcher I've actually seen this happen on ASP.NET MVC5 as well. I have a suspicion I've found the cause, but if that's not it I'll try that next.

**Comment by muratg** [comment]
cc @pakrym

**Comment by pakrym** [comment]
@kanadaj do you see other threads being blocked on `Monitor.Enter`?

**Comment by kanadaj** [comment]
@pakrym Yes, there are another 72 other threads all waiting on Thread 151 on the same lock
![image](https://user-images.githubusercontent.com/3247357/48649050-42abb080-e9e9-11e8-9c83-1c3d5d2ecb2e.png)

**Comment by kanadaj** [comment]
Also, it seems like removing the non-thrown exception from our old custom Sentry RavenSharp middleware  seems to have removed the issue.

In essence, the middleware did something like this:
```
var request = context.Request;
var body = request.Body;
try
{
    request.EnableRewind();
    await _next(context);
}
catch(Exception ex)
{
    if (ex is OperationCanceledException || ex is Win32Exception)
    {
        // Let's ignore these errors since they are meaningles
        throw;
    }
    if(ex is DbEntityValidationException dbex)
    {
        var errorMessages = dbex.EntityValidationErrors.Select(x => $"{x.Entry.Entity.GetType().Name}.{string.Join("; ", x.ValidationErrors.Select(y => $"{y.PropertyName}: {y.ErrorMessage}"))}");
        var err = new Exception($"Database Validation error: {string.Join("; ", errorMessages)}", ex);
        await _ravenClient.CaptureAsync(ex);
    }
    throw;
}```
So maybe it's related to creating an exception without throwing? Could that cause a throw inside the global Exception lock?

**Comment by pakrym** [comment]
This is all very strange. I don't see how anything might be hanging while holding the lock in question (https://referencesource.microsoft.com/#mscorlib/system/exception.cs,837).

**Comment by pakrym** [comment]
It might be a side effect of attaching too many stacks to the same exception object because of aspnet/KestrelHttpServer#2806

**Comment by kanadaj** [comment]
@pakrym Me neither, but it somehow happens. I have 2 possible scenarios in mind, 1 is that somehow something in that lock throws... But that's rather unlikely, since I believe the stacktrace would show a recursion there, which it doesn't. The other guess is that maybe something externally hangs at that point, like the Application Insights or Retrace - we've observed the problem surfacing very rarely without either, but with that middleware added, Application Insights' profiler causes a hang _very_ quickly, and Retrace's own agent causes hangs after a couple hours. I have yet to try the code without the middleware with Application Insights' profiler again.

I've also seen identical issues without any profiler attached (though I have no dumps from that, so no guarantee it's the same, but all external symptoms are completely identical, so likely) but far less often, maybe once every few months. It's all very weird, and even diagnosing this far was hard since we've looked for the problem in other places. The error messages collected by Serilog and Sentry point all over the place, usually reporting timeouts from Redis and Entity Framework 6. But in hindsight that's only a consequence of the deadlock, since first chance exceptions would also hang, which would look like a timeout to Entity Framework.

Other things that usually happen when the problem surfaces is a massive spike in open connections - I recon it's maybe Kestrel trying to add more threads since the ones hanging can't serve requests - and increased CPU load on the victim server, though usually only to around 80% on a V2 4 core Azure Web Service instance.

**Comment by kanadaj** [comment]
> It might be a side effect of attaching too many stacks to the same exception object because of [aspnet/KestrelHttpServer#2806](https://github.com/aspnet/KestrelHttpServer/issues/2806)

I've seen identical issues on ASP.NET MVC 5 before we migrated from it. I don't think that part of the pipeline is shared?

**Comment by kanadaj** [comment]
I only have a stacktrace for the inner exception:
```
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitable.GetResult() in /_/src/Kestrel.Transport.Sockets/Internal/SocketAwaitable.cs:line 40
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.<ProcessReceives>d__25.MoveNext() in /_/src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs:line 164
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in f:\dd\ndp\clr\src\BCL\system\runtime\exceptionservices\exceptionservicescommon.cs:line 133
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) in f:\dd\ndp\clr\src\BCL\system\runtime\compilerservices\TaskAwaiter.cs:line 156
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.<DoReceive>d__24.MoveNext() in /_/src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs:line 110
```

_stackTraceCopy is null, as are all the copied values of exceptionDispatchInfo.

**Comment by kanadaj** [comment]
![image](https://user-images.githubusercontent.com/3247357/48650818-b9e44300-e9ef-11e8-8632-2528fdb4f093.png)

**Comment by kanadaj** [comment]
> Throwing in the lock would release the lock so I doubt that's it.

Wouldn't the throw have to finish before the lock is released?
 
> Pipelines didn't exist until AspNetCore 2.0.

That's what I guessed. Sadly there are so many external factors in play here (from profilers to whatever things Azure App Services run for monitoring) it's very hard to pinpoint why this is happening. None of this really makes any sense to me. It makes even less sense to me that removing the middleware seems to have fixed the problem.

**Comment by kanadaj** [comment]
The reason I suspect some sort of recursion is that the thread seems to be waiting to enter a lock that's already owned by itself

**Comment by anurse** [comment]
This issue appears to have gotten stale. If you're still having a problem with this, please feel free to chime in and we can try to resume investigating :)
